site-src/geps/gep-1867.md

```diff
@@ -0,0 +1,137 @@
+
+Some parallels in existing APIs:
+
+[Policy Attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment) offers a hierarchy of defaults and overrides, allowing attachment to GatewayClass and Gateway.
+This is similar to our needs here, but representing infrastructure configuration as a "Policy" is a bit problematic, and the existing mechanisms have no hierarchy.
+
+In core Kubernetes, Pods declare their requirements (for example, CPU requests) inline in the Pod resource; there is not a `ResourceClass` API that abstracts these further.
+These higher level abstractions are handled by layered APIs (whether this is a CRD, an admission webhook, CI/CD tooling, etc).
+This allows users the flexibility to easily configure things per-pod basis.
+If the infrastructure admin wants to impose defaults or requirements on this flexibility, they are able to do so (in fact, `LimitRanger` provides a built in mechanism to do so).
+
+### Dynamic Changes
+
+Currently, `GatewayClass` is intended as a *template*.
+Changes to it are not expected to change deployed `Gateway`s.
+
+This makes usage problematic in a declarative way.
+For example, if I wanted to represent a `version` field and change that to trigger an upgrade, I would need to create an entirely new `Gateway`.
+
+
+
+## API
+
+In order to address the concerns above, I propose a standard `infrastructure` API is added to `Gateway` and `GatewayClass`.
+Note the important part of this is the `Gateway` change; the `GatewayClass` aspect is mostly for consistency.
+
+Note that many of the fields under `GatewayInfrastructure` refer to WIP GEPs.
+These fields are not tied to this GEP, and will only be included if the respective GEPs are approved.
+However, they are shown below to give a sense of the use cases the field can support.
+
+```go
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+	Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+
+
+type GatewayClassSpec struct {
+  // Infrastructure defines infrastructure level attributes for all Gateways in this class.
+  // A Gateway may provide configuration for the same values; as all fields in GatewayInfrastructure are implementation specific,
+  // the merging logic between these is as well.
+	Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+
+
+type GatewayInfrastructure struct {
+  // AttachTo marks this Gateway as a child to the referenced parent Gateway.
+  // See GEP-1713 for details, which this depends on.
+  AttachTo LocalObjectReference
+
+  // Scope defines the "scope" of the Gateway's addresses.
+  // Valid options are ClusterLocal or Public
+  // See GEP-1651 for details, which this depends on.
+  // NOTE: that GEP is likely to be implemented in a different way, at which point this field can be removed from this struct.
+  Scope AddressScope
```

Comment:
If we do it in spec.addresses we should not do it in spec.infra. I meant to indicate that in my comment that we may remove this depending on how 1651 goes

Reply:
I actually think I like it in `infra` since in many/most cases only a single address scope is going to be supported by a Gateway. Are there important use cases I'm missing where we'd want to be able to request multiple address scopes?

Reply:
Statically assign a LoadBalancer IP and a ClusterIP, I guess? not sure why you would actually do that, though

Probably a discussion for 1651

Reply:
Yes, I think this raises the point that #1651 implies that it will be useful to have one Gateway with multiple address scopes.

Personally, I think it may be better to encourage users to have different Gateways with different scopes, to make decisions about where to expose things more explicit (you'll need to point your Routes at _both_ Gateways if you want your route available in both scopes).